src/url.cpp

```diff
@@ -457,4 +457,54 @@ namespace Url
         str.resize(dest);
     }
 
+    Url& Url::deparam(const std::unordered_set<std::string>& blacklist)
+    {
+        remove_params(params_, blacklist, ';');
+        remove_params(query_, blacklist, '&');
+        return *this;
+    }
+
+    void Url::remove_params(std::string& str, const std::unordered_set<std::string>& blacklist, const char separator)
+    {
+
+        std::vector<std::string> pieces;
+        size_t previous = 0;
+        for (size_t index = str.find(separator)
+            ; index != std::string::npos
+            ; previous = index + 1, index = str.find(separator, previous))
+        {
+            std::string piece = str.substr(previous, index - previous);
+            std::string name = piece.substr(0, piece.find('='));
+            std::transform(name.begin(), name.end(), name.begin(), ::tolower);
+
+            if (blacklist.find(name) == blacklist.end())
+            {
+                pieces.push_back(piece);
+            }
+        }
+
+        if (previous < str.length())
+        {
+            std::string piece = str.substr(previous);
+            std::string name = piece.substr(0, piece.find('='));
+            std::transform(name.begin(), name.end(), name.begin(), ::tolower);
```

Comment:
I'm pretty sure that `::tolower` will only do the right thing for the ASCII range. Isn't this an issue if you have UTF-8 encoded strings? Furthermore, this is probably an issue throughout where you are doing "case-insensitive" comparisons.